src/lib/schematics/ng-update/upgrade-rules/package-imports-v8/updateAngularMaterialImportsRule.ts

```diff
@@ -0,0 +1,172 @@
+    typescriptOnly: true,
+  };
+
+  static ONLY_SUBPACKAGE_FAILURE_STR = Lint.Utils.dedent`
+      Importing from @angular/material is deprecated. Instead import from
+      subpackage the symbol belongs to. e.g. import {MatButtonModule} from
+      '@angular/material/button'`;
+  static NO_IMPORT_NAMED_SYMBOLS_FAILURE_STR = Lint.Utils.dedent`
+      Imports from Angular Material should import specific symbols rather than
+      importing the entire Angular Material library.`;
+  static SYMBOL_NOT_FOUND_FAILURE_STR = ` was not found in the Material library.`;
+
+  applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+    return this.applyWithFunction(sourceFile, walk, true, program.getTypeChecker());
+  }
+}
+
+/**
+ * A walker to walk a given source file to check for imports from the
+ * @angular/material module.
+ */
+function walk(ctx: Lint.WalkContext<boolean>, checker: ts.TypeChecker): void {
+  // The source file to walk.
+  const sf = ctx.sourceFile;
+  const cb = (declaration: ts.Node) => {
+    // Only look at import declarations.
+    if (!ts.isImportDeclaration(declaration)) {
+      return;
+    }
+    const importLocation = declaration.moduleSpecifier.getText(sf);
+    // If the import module is not @angular/material, skip check.
+    if (importLocation !== ANGULAR_MATERIAL_IMPORT_PATH) {
+      return;
+    }
+
+    // If no import clause is found, or nothing is named as a binding in the
+    // import, add failure saying to import symbols in clause.
+    if (!declaration.importClause || !declaration.importClause.namedBindings) {
+      return ctx.addFailureAtNode(declaration, Rule.NO_IMPORT_NAMED_SYMBOLS_FAILURE_STR);
+    }
+
+    // All named bindings in import clauses must be named symbols, otherwise add
+    // failure saying to import symbols in clause.
+    if (!ts.isNamedImports(declaration.importClause.namedBindings)) {
+      return ctx.addFailureAtNode(declaration, Rule.NO_IMPORT_NAMED_SYMBOLS_FAILURE_STR);
+    }
+
+    // If no symbols are in the named bindings then add failure saying to
+    // import symbols in clause.
+    if (!declaration.importClause.namedBindings.elements.length) {
+      return ctx.addFailureAtNode(declaration, Rule.NO_IMPORT_NAMED_SYMBOLS_FAILURE_STR);
+    }
+
+    // Map of submodule locations to arrays of imported symbols.
+    const importMap = new Map<string, Set<string>>();
+
+    // Determine the subpackage each symbol in the namedBinding comes from.
+    for (const importedSymbol of declaration.importClause.namedBindings.elements) {
+      // Confirm the named import is a symbol that can be looked up.
+      if (!ts.isIdentifier(importedSymbol.name)) {
+        return ctx.addFailureAtNode(
+            importedSymbol, importedSymbol.getFullText(sf) + Rule.SYMBOL_NOT_FOUND_FAILURE_STR);
+      }
+      // Get the original symbol where it is declared upstream.
+      const symbol = getOriginalSymbol(importedSymbol.name, checker);
```

Comment:
The easiest way to get the original value declaration is using:

```ts
const type= this.getTypeChecker().getTypeAtLocation(importedSymbol.name);
const symbol = type.getSymbol();

if (!symbol || !symbol.valueDeclaration) {
  // skip
}

const decl = symbol.valueDeclaration;
const sourceFile = decl.getSourceFile();
```

<sup>Avoids the use of tsutils; and also is more idiomatic as we want the value declaration of the imported type</sup>